File: 다이나믹 프로그래밍/2096 내려가기/2096.py
import sys

input = sys.stdin.readline

N = int(input())

dp_max = [[0] * 3 for _ in range(2)]
dp_min = [[0] * 3 for _ in range(2)]

for i in range(N):
    a, b, c = map(int, input().split())
    
    if i == 0:
        dp_max[0][0], dp_max[0][1], dp_max[0][2] = a, b, c
        dp_min[0][0], dp_min[0][1], dp_min[0][2] = a, b, c
    else:
        dp_max[1][0] = max(dp_max[0][0] + a, dp_max[0][1] + a)
        dp_min[1][0] = min(dp_min[0][0] + a, dp_min[0][1] + a)

        dp_max[1][1] = max(dp_max[0][0] + b, dp_max[0][1] + b, dp_max[0][2] + b)
        dp_min[1][1] = min(dp_min[0][0] + b, dp_min[0][1] + b, dp_min[0][2] + b)

        dp_max[1][2] = max(dp_max[0][1] + c, dp_max[0][2] + c)
        dp_min[1][2] = min(dp_min[0][1] + c, dp_min[0][2] + c)

        dp_max[0][0], dp_max[0][1], dp_max[0][2] = dp_max[1][0], dp_max[1][1], dp_max[1][2]
        dp_min[0][0], dp_min[0][1], dp_min[0][2] = dp_min[1][0], dp_min[1][1], dp_min[1][2]

print(max(dp_max[0]))
print(min(dp_min[0]))

File: 다이나믹 프로그래밍/2096 내려가기/2096.js
let fs = require('fs');
let filePath = process.platform === 'linux' ? '/dev/stdin' : '/input.txt';
let input = fs
  .readFileSync(__dirname + filePath)
  .toString()
  .trim()
  .split('\n');

const N = +input.shift();
let board = [];
for (let i = 0; i < N; i++) {
  board[i] = input[i].split(' ').map(Number);
}

let dp_max = [...new Array(2)].map((_) => Array(3).fill(0));
let dp_min = [...new Array(2)].map((_) => Array(3).fill(0));

dp_max[0][0] = board[0][0];
dp_max[0][1] = board[0][1];
dp_max[0][2] = board[0][2];

dp_min[0][0] = board[0][0];
dp_min[0][1] = board[0][1];
dp_min[0][2] = board[0][2];

for (let i = 2; i <= N; i++) {
  dp_max[1][0] = Math.max(dp_max[0][0] + board[i - 1][0], dp_max[0][1] + board[i - 1][0]);
  dp_min[1][0] = Math.min(dp_min[0][0] + board[i - 1][0], dp_min[0][1] + board[i - 1][0]);

  dp_max[1][1] = Math.max(
    dp_max[0][0] + board[i - 1][1],
    dp_max[0][1] + board[i - 1][1],
    dp_max[0][2] + board[i - 1][1]
  );
  dp_min[1][1] = Math.min(
    dp_min[0][0] + board[i - 1][1],
    dp_min[0][1] + board[i - 1][1],
    dp_min[0][2] + board[i - 1][1]
  );

  dp_max[1][2] = Math.max(dp_max[0][1] + board[i - 1][2], dp_max[0][2] + board[i - 1][2]);
  dp_min[1][2] = Math.min(dp_min[0][1] + board[i - 1][2], dp_min[0][2] + board[i - 1][2]);

  dp_max[0][0] = dp_max[1][0];
  dp_max[0][1] = dp_max[1][1];
  dp_max[0][2] = dp_max[1][2];

  dp_min[0][0] = dp_min[1][0];
  dp_min[0][1] = dp_min[1][1];
  dp_min[0][2] = dp_min[1][2];
}

console.log(Math.max(...dp_max[1]));
console.log(Math.min(...dp_min[1]));
